### PR TITLE
Feature/66 List custom charts in the Airports app

### PR DESCRIPTION
**Pull request by oobayly**
See #66 

The new LocalFile API looks for any PDF or image files in the `charts/{ICAO}` directory, and adds them as a new "Other" category. The new files are given a prefix of `LF-{index}`.

Changes made:
- Add "Other" chart type
- Change `ChartService` constructor to only pass the program path instead of navigraph path.
- Add LocalFile chart API

Double check the following, I've added comments to the relevant changes:
- [ ] Method for checking if the file is a PDF. It's nominally case insensitive, but won't match XYZ.PdF
- [ ] Method for loading the image. I cobbled this together, but even though it works, I'm not convinced it's the best method
- [ ] I've probably missed out some error handling

Screenshots:
image
image
image

### CODE REVIEW

File: src/avitab/AviTab.cpp [hunk 1]
```diff
@@ -35,7 +35,7 @@ AviTab::AviTab(std::shared_ptr<Environment> environment):
     if (env->getConfig()->getBool("/AviTab/loadNavData")) {
         env->loadNavWorldInBackground();
     }
-    chartService = std::make_shared<apis::ChartService>(env->getProgramPath() + "/Navigraph/");
+    chartService = std::make_shared<apis::ChartService>(env->getProgramPath());
```
Comment (oobayly) [inline src/avitab/AviTab.cpp:38]: As I don't have access to the Navigraph build, I can't check whether I've broken anything. I don't think I have, but...

File: src/charts/liblocalfile/LocalFileAPI.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        }
+    } catch (const std::exception &e) {
+        logger::verbose(e.what());
+    }
+
+    return charts;
+}
+
+void LocalFileAPI::loadChart(std::shared_ptr<LocalFileChart> chart) {
+    std::ifstream instream(chart->getPath(), std::ios::in | std::ios::binary);
```
Comment (fpw) [inline src/charts/liblocalfile/LocalFileAPI.cpp:67]: The fstream API is highly incompatible on Windows - it expects ANSI paths and we're using UTF-8 everywhere. I added a compatible drop-in replacement that chooses a working implementating on all supported platforms. Please replace std::*stream by fs::*stream from platform.h

**Review by fpw — CHANGES_REQUESTED** [review]

**Comment by fpw** [comment]
Looking great, still testing a bit - will likely merge it soon!

File: src/avitab/AviTab.cpp (hunk 1 above)
Comment (fpw) [reply src/avitab/AviTab.cpp:38]: Checked, works.

**Comment by fpw** [comment]
Did a little change in the approach: The "data attach" feature of the Chart class was supposed to be used for storing temporarily downloaded data that's not allowed to be cached on disk, mainly for conforming to chart providers terms of services.

Here, the file is already available on disk, so we don't need the attach feature. This saves memory because the TileSource class will create a memory-cache of the visible portions anyways.

And by simply loading the file from disk, we can always use PDFSource since that can also render many image formats (through mupdf). 

Thanks again, the approach of adding local charts as a chart provider was a great idea!

**Comment by oobayly** [comment]
> Did a little change in the approach: The "data attach" feature of the Chart class was supposed to be used for storing temporarily downloaded data that's not allowed to be cached on disk, mainly for conforming to chart providers terms of services.
> 
> Here, the file is already available on disk, so we don't need the attach feature. This saves memory because the TileSource class will create a memory-cache of the visible portions anyways.
> 
> And by simply loading the file from disk, we can always use PDFSource since that can also render many image formats (through mupdf).
> 
> Thanks again, the approach of adding local charts as a chart provider was a great idea!

That makes a lot more sense. It did feel a little strange caching the files in memory, but I couldn't quite get my head around how to pass the data into the viewer component so I resorted to copy-pasta from other components!

**Pull request merged**